x/nodes/keeper/reward.go

```diff
@@ -14,19 +14,27 @@ func (k Keeper) RewardForRelays(ctx sdk.Ctx, relays sdk.BigInt, address sdk.Addr
 	// feature flags
 	isAfterRSCAL := k.Cdc.IsAfterNamedFeatureActivationHeight(ctx.BlockHeight(), codec.RSCALKey)
 	isNonCustodialActive := k.Cdc.IsAfterNonCustodialUpgrade(ctx.BlockHeight())
+	isEnforceMaxChains := k.Cdc.IsAfterEnforceMaxChainsUpgrade(ctx.BlockHeight())
 
 	// The conditions of the original non-custodial issue:
 	isDuringFirstNonCustodialIssue := isNonCustodialActive && isAfterRSCAL && ctx.BlockHeight() <= codec.NonCustodial1RollbackHeight
 
 	// get validator
 	validator, found := k.GetValidator(ctx, address)
 
-	//adding "&& (isAfterRSCAL || isAfterNonCustodial)" to sync from scratch as weighted stake and non-custodial introduced this requirement
+	// adding "&& (isAfterRSCAL || isAfterNonCustodial)" to sync from scratch as weighted stake and non-custodial introduced this requirement
 	if !found && (isAfterRSCAL || isNonCustodialActive) {
 		ctx.Logger().Error(fmt.Errorf("no validator found for address %s; at height %d\n", address.String(), ctx.BlockHeight()).Error())
 		return sdk.ZeroInt()
 	}
 
+	// check if the node is overstaked for max chains and if so  and if so dont reward them
+	// this ensures that nodes can still participate in the network but are not rewarded for relays
+	// if they are overstaked for max chains, ensuring that the network functions properly
+	if isEnforceMaxChains && int64(len(validator.GetChains())) > k.GetNodeMaxChains(ctx) {
```

Comment:
@msmania So i removed the check to select nodes for a session and replaced it with a check in the rewarding of the node for a session. If they are overstaked they get no rewards. Is this more in line with your idea?